[BugFix] Json loader will lead to BE crash if it reads an error row.

xiaoyong-z opened the pull request:
Signed-off-by: xyz <a997647204@gmail.com>

## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #6052
Fixes #6051
Fixes #6098
Fixes #6097
## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->

Now, the JSON loader will lead to BE crash if it meets an error row.
In the non-strict model, the data(number, string) which cannot be parsed by simdjson will be filled with NULL. However, in the strict model, the data which can not be parsed by simdjson will not be filled with NULL. This will lead to the error column containing one row less than the other columns. In this pr, we will remove the error row before we continue to parse other rows.

Apart from this, there's a bug in simdjson which will recognize -9223372036854775808 as an unsigned number, this pr also adds additional logic to handle this.

rickif commented on be/src/formats/json/numeric_column.cpp:
```diff
@@ -39,10 +39,30 @@ static Status add_column_with_numeric_value(FixedLengthColumn<T>* column, const
+        // This is because currently there exists a bug in simdjson, for number <= -9223372036854775808,
+        // simdjson will recognized it as unsigned_integer, and use get_uint64() to parse it, which will
+        // raise an exception here. So here we need to recheck by get_int64().
+        // This bug in simdjson is expected to be solved by the following commit, lator we can remove
```
`lator we can remove` - > `later we can remove`

rickif commented on be/src/exec/vectorized/json_scanner.cpp:
```diff
@@ -437,6 +437,15 @@ Status JsonReader::_read_rows(Chunk* chunk, int32_t rows_to_read, int32_t* rows_
                 _state->append_error_msg_to_file(std::string(sv.data(), sv.size()), st.to_string());
                 LOG(WARNING) << "failed to construct row: " << st;
             }
+            if (st.is_data_quality_error() || st.is_invalid_argument()) {
```
It seems that `read_chunk()` at line 405 has handled this but dose not work. Could you please check it and retain one of them?

rickif commented on be/src/formats/json/numeric_column.cpp:
```diff
@@ -39,10 +39,30 @@ static Status add_column_with_numeric_value(FixedLengthColumn<T>* column, const
+            if (value->get_int64() == std::numeric_limits<int64_t>::min()) {
+                checked_cast_flag = checked_cast(std::numeric_limits<int64_t>::min(), &out);
+            } else {
+                throw simdjson::simdjson_error(simdjson::error_code::NUMBER_OUT_OF_RANGE);
```
Is it better to return an error instead?

xiaoyong-z replied:
In the original code, it should raise an exception here because it can not be parsed by simdjson get_uint64(). 
Here we recheck if it is a std::numeric_limits<int64_t>::min() due to a bug in simdjson. If it is not, i think we should still throw an exception here?

xiaoyong-z committed:
[BugFix] json loader will lead to BE if it reads an error row.

Now, the json loader will lead to BE crash if it meets an error row.
In the non strict model, the data(number, string) which cannot be parsed by simdjson will be filled with NULL. However, in the strict model, the data which can not be parsed by simdjson will not be filled with NULL. This will lead to the error column containing one row less than other columns. In this pr, we will remove the error row before we continue to parse other rows.

Apart from this, there's a bug in simdjson which will recognize -9223372036854775808 as unsigned number, this pr also adds additional logic to handle this.

Signed-off-by: xyz <a997647204@gmail.com>

xiaoyong-z committed:
1. catch unexpected exception
2. append null for data_quality_error under non strict model

Signed-off-by: xyz <a997647204@gmail.com>

xiaoyong-z committed:
return error when parsing invalid key

Signed-off-by: xyz <a997647204@gmail.com>

mofeiatwork commented on be/src/exec/vectorized/json_scanner.cpp:
```diff
@@ -73,9 +73,21 @@ StatusOr<ChunkPtr> JsonScanner::get_next() {
+    try {
+        status = _cur_file_reader->read_chunk(src_chunk.get(), _max_chunk_size);
+    } catch (simdjson::simdjson_error& e) {
+        auto err_msg = strings::Substitute("The format of the json file may be wrong, stop json loader.");
```
Unrecognized json format?

xiaoyong-z committed:
modify error message

Signed-off-by: xyz <a997647204@gmail.com>

rickif commented on be/src/exec/vectorized/json_scanner.cpp:
```diff
@@ -73,9 +73,21 @@ StatusOr<ChunkPtr> JsonScanner::get_next() {
+    try {
+        status = _cur_file_reader->read_chunk(src_chunk.get(), _max_chunk_size);
+    } catch (simdjson::simdjson_error& e) {
+        auto err_msg = strings::Substitute("Unrecognized json format, stop json loader.");
```
`strings::Substitute` is unnecessary here since no format arg is provided.

xiaoyong-z commented on be/src/formats/json/nullable_column.cpp:
```diff
@@ -185,7 +185,7 @@ Status add_nullable_column(Column* column, const TypeDescriptor& type_desc, cons
 
         auto st = add_nullable_column(column, type_desc, name, value);
-        if (st.is_invalid_argument() && invalid_as_null) {
+        if ((st.is_invalid_argument() || st.is_data_quality_error()) && invalid_as_null) {
```
may be `!st.ok()` ?

xiaoyong-z committed:
modify (st.is_invalid_argument() || st.is_data_quality_error()) to !st.ok()

Signed-off-by: xyz <a997647204@gmail.com>
